ddtrace/contrib/kafka/patch.py

```diff
@@ -0,0 +1,125 @@
+from ddtrace.constants import SPAN_KIND
+from ddtrace.constants import SPAN_MEASURED_KEY
+from ddtrace.contrib import trace_utils
+from ddtrace.ext import SpanKind
+from ddtrace.ext import SpanTypes
+from ddtrace.ext import kafka as kafkax
+from ddtrace.internal.compat import ensure_text
+from ddtrace.internal.constants import COMPONENT
+from ddtrace.internal.constants import MESSAGING_SYSTEM
+from ddtrace.internal.utils import ArgumentError
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.pin import Pin
+from ddtrace.vendor.wrapt import ObjectProxy
+
+
+_Producer = confluent_kafka.Producer
+_Consumer = confluent_kafka.Consumer
+
+
+config._add(
+    "kafka",
+    dict(_default_service="kafka"),
+)
+
+
+class TracedProducer(ObjectProxy):
+    def __init__(self, *args, **kwargs):
+        producer = _Producer(*args, **kwargs)
+        super(TracedProducer, self).__init__(producer)
+        Pin().onto(self)
+
+    def produce(self, *args, **kwargs):
+        func = self.__wrapped__.produce
+        topic = get_argument_value(args, kwargs, 0, "topic")
+        try:
+            value = get_argument_value(args, kwargs, 1, "value")
+        except ArgumentError:
+            value = None
+        message_key = kwargs.get("key", "")
+        partition = kwargs.get("partition", -1)
+
+        pin = Pin.get_from(self)
+        if not pin or not pin.enabled():
+            return func(*args, **kwargs)
+
+        with pin.tracer.trace(
+            kafkax.PRODUCE,
+            service=trace_utils.ext_service(pin, config.kafka),
+            span_type=SpanTypes.WORKER,
+        ) as span:
+            span.set_tag_str(MESSAGING_SYSTEM, kafkax.SERVICE)
+            span.set_tag_str(COMPONENT, config.kafka.integration_name)
+            span.set_tag_str(SPAN_KIND, SpanKind.PRODUCER)
+            span.set_tag_str(kafkax.TOPIC, topic)
+            span.set_tag_str(kafkax.MESSAGE_KEY, ensure_text(message_key))
+            span.set_tag(kafkax.PARTITION, partition)
+            span.set_tag_str(kafkax.TOMBSTONE, str(value is None))
+            span.set_tag(SPAN_MEASURED_KEY)
+            rate = config.kafka.get_analytics_sample_rate()
+            if rate is not None:
+                span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, rate)
+            return func(*args, **kwargs)
+
+    # in older versions of confluent_kafka, bool(Producer()) evaluates to False,
+    # which makes the Pin functionality ignore it.
+    def __bool__(self):
+        return True
+
+    __nonzero__ = __bool__
+
+
+class TracedConsumer(ObjectProxy):
+    def __init__(self, *args, **kwargs):
+        consumer = _Consumer(*args, **kwargs)
+        super(TracedConsumer, self).__init__(consumer)
+        Pin().onto(self)
+
+    def poll(self, *args, **kwargs):
+        func = self.__wrapped__.poll
+        pin = Pin.get_from(self)
+        if not pin or not pin.enabled():
+            return func(*args, **kwargs)
+
+        with pin.tracer.trace(
```

Comment:
so usually, the code is:
```
message = consumer.consume()
doSomething(message)
doSomethingElse(message)
```
what would be nice (and what .NET allows for for eg) is to keep the span alive until the next call of consumer.consume(). That way, the span from consume is a parent span of `doSomething()` and `doSomethingElse` in my example. Is that something that would be possible to do?

Reply:
@piochelepiotr this PR instruments `Consumer.poll()`, which is similar but not exactly the same as `consume()` AFAIK. Should we wrap `Consumer.consume()` instead? 

I'm not opposed to keeping the span alive, but my initial worry is for edge cases where `poll()` is called but no subsequent `poll()` is called. How do we know to finish the span without the next call of `consumer.poll()`?

Reply:
@Yun-Kim I believe `[consume](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Consumer.consume)()` is a convenience wrapper around `poll()` that returns a given number of messages rather than `poll()`'s 1-or-0 return guarantee.

It seems to me that we should not keep `poll()` spans alive because of the issue you mentioned. Separately, we could add instrumentation for `consume()` that closed spans either on the next `consume()` call or on `consume()` timeout. I'd personally be fine with putting that in a subsequent pull request.

Reply:
@emmettbutler agreed, I think this behavior doesn't need to block this PR and can be implemented in a future PR after spec-ing out the behavior more thoroughly.